### DISCUSSION TITLE
CPU 占用高好像不是 bug，是真的高

### DISCUSSION THREAD

**Discussion by Xm798** [Arondight/Adachi-BOT#668] [General]
```
[yakult@centos ~]$ top -bn 1 -i -c
top - 14:48:41 up  3:32,  2 users,  load average: 2.18, 1.24, 0.57
Tasks:  98 total,   2 running,  96 sleeping,   0 stopped,   0 zombie
%Cpu(s): 72.2 us, 27.8 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   1826.7 total,    264.4 free,    452.7 used,   1109.6 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   1201.9 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  31056 yakult    20   0   24.5g  97136  82728 R  58.8   5.2   0:13.58 /usr/lib64/chromium-browser/chromium-brows+
  31032 yakult    20   0   16.6g  88224  75516 S  35.3   4.7   0:07.43 /usr/lib64/chromium-browser/chromium-brows+
```

直接拉满，离谱

**Comment by Xm798** [comment]
记得之前 chrome 占用貌似没这么高的

**Comment by mark9804** [comment]
不会吧……？我这边不绘图的时候 chromium 动都懒得动，绘图的时候才能看到高占用
<img width="409" alt="image" src="https://user-images.githubusercontent.com/9006264/155078700-8bbbbfb1-58fd-407f-8097-752b6fc2f021.png">

平时 chromium 都是 0.0%，树莓派 4B+ ，感觉一般不至于能有比树莓派 CPU 更烂的 VPS……

**Reply by Xm798** [reply]
你也用servercat

**Reply by mark9804** [reply]
> 你也用servercat

没办法，是真的好使，就是作者一直不出macOS版本的，我比他还急

**Comment by Xm798** [comment]
我这为什么 chrome 的 CPU 一直都是拉满的……
阿里云的 1c2g
```
[yakult@centos Adachi-BOT-Data]$ top
top - 15:04:45 up  3:48,  2 users,  load average: 1.61, 1.51, 1.19
Tasks:  98 total,   2 running,  96 sleeping,   0 stopped,   0 zombie
%Cpu(s): 76.7 us, 22.6 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.7 hi,  0.0 si,  0.0 st
MiB Mem :   1826.7 total,    186.9 free,    497.0 used,   1142.9 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   1185.2 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                    
  31056 yakult    20   0   24.5g  97452  83560 R  64.3   5.2  10:28.62 chromium-browse                            
  31032 yakult    20   0   16.6g  88976  76116 S  35.0   4.8   5:42.23 chromium-browse                            
  28983 root      20   0  715224  20480  10780 S   0.7   1.1   0:09.44 nezha-agent                                
  30988 yakult    20   0   10.9g 145984  36768 S   0.3   7.8   0:05.31 node /home/yaku                            
  31201 yakult    20   0   64480   4772   4020 R   0.3   0.3   0:00.02 top
```

**Reply by Xm798** [reply]
![image](https://user-images.githubusercontent.com/48074861/155080199-bb999bca-f3be-43f9-a04a-16432d9719c7.png)

**Comment by mark9804** [comment]
我不能复现，好几把怪

会不会是你的 chromium 源有问题？我感觉这种情况像是 `--no-zygote` 没有生效，但是正常情况下不应该出现这个样子的

**Reply by mark9804** [reply]
```
chromium-browser --version
apt policy chromium-browser
```

**Reply by Xm798** [reply]
> 我不能复现，好几把怪
> 
> 会不会是你的 chromium 源有问题？我感觉这种情况像是 `--no-zygote` 没有生效，但是正常情况下不应该出现这个样子的

Chromium 96.0.4664.110 Fedora Project

我系统是 Centos-8.2

**Reply by Xm798** [reply]
我倒是一直在 Centos 跑的，但是记得之前占用没这么高

**Reply by Arondight** [reply]
```
平台：linux（CentOS Stream）
内核：4.18.0-365.el8.x86_64
架构：x64
CPU：10.80%（Intel Xeon® Platinum 8255C @ 2.5Ghz）
内存：21.37%（836 MB / 3.91 GB）
启动：8 days
数据：8.75 MB
```

我环境跟你差不多，没这个问题， htop 好几页才能看到浏览器

**Reply by Xm798** [reply]
> LSB Version:	:core-4.1-amd64:core-4.1-noarch
> Distributor ID:	CentOS
> Description:	CentOS Linux release 8.2.2004 (Core) 
> Release:	8.2.2004
> Codename:	Core

好特么奇怪，我之前在 synologo 的 vmm 虚拟机上跑的，吃 nas 主机 cpu 太厉害，丢到阿里云上了，两边都吃满 CPU，太奇怪了吧。。。

阿里云这个从系统到项目全都是今天重装的

![image](https://user-images.githubusercontent.com/48074861/155090075-e6cf181d-06b7-43e3-a9ee-31fed45cfcca.png)

**Reply by mark9804** [reply]
你跑一下下面这行看一下是哪来的chrome吧，感觉源有问题的可能性稍微大一些

```
apt policy chromium-browser

```

**Reply by Xm798** [reply]
我就是 yum install chromium 安装的，apt 在 Centos 上用不了

**Reply by Arondight** [reply]
你在虚拟机里面手动拉一个无头起来看看，可以改改 tools/view.js 随便用一个页面把无头拉起来，看看占用高么

**Comment by Xm798** [comment]
过了一晚自己好了……真 tm 玄学啊，反应也变快了，之前绘图要卡好多秒

![image](https://user-images.githubusercontent.com/48074861/155255507-5e6a176e-f9e2-4265-aeab-58d157963b32.png)

**Reply by Arondight** [reply]
有没有可能是你的 vps 超分太多了，恰巧这个节点有几个计算大户

**Reply by Xm798** [reply]
应该不会，之前在本地也是这样的，持续了好几天，吃的我 NAS 的 CPU 一直飘红，我才放到闲置的阿里云上去的

**Reply by Arondight** [reply]
要不切到stream试试 #536

**Reply by mark9804** [reply]
会不会某个版本的 chromium 有这个问题

**Reply by Xm798** [reply]
> 会不会某个版本的 chromium 有这个问题

不清楚耶，但现在它好了，什么都没做，就很奇怪

**Reply by Xm798** [reply]
> 要不切到stream试试 #536 

再出问题就切过去试试，提前停止支持真的麻烦

**Comment by mark9804** [comment]
额，我昨天把 chromium 从 95 更新到 98 之后也遇到了类似的情况， chromium 被拉起后持续空载占用大量 CPU ，设备是树莓派，64位 raspbian，兼容性可能不是很好，不确定是不是这个原因

```bash
> chromium-browser --version
< Chromium 98.0.4758.106 Built on Debian , running on Debian 11
```

```
平台：linux（Debian GNU/Linux）
内核：5.10.92-v8+
架构：arm64
CPU：1.34%（Sony UK BCM2711 @ 1.8Ghz）
内存：31.61%（1.26 GB / 3.98 GB）
启动：a day
数据：800 kB

PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Reply by mark9804** [reply]
今天树莓派上面能够稳定复现了，但是我的 mac 上面复现不了
Mac 上的 chromium 版本号为 `Version 99.0.4844.0 (Developer Build) (x86_64)`

空载运行表现截图如图
![EE5056CC-E130-4A5E-8F09-181803FEA923_1_102_o](https://user-images.githubusercontent.com/9006264/155865828-4b043d5e-c5f5-4c3c-9715-f7d32c123907.jpeg)

**Reply by Xm798** [reply]
我这边版本是 Chromium 96.0.4664.110 Fedora Project
就自从那天莫名其妙好了之后到目前为止还是好的

**Reply by mark9804** [reply]
> 就自从那天莫名其妙好了之后到目前为止还是好的

我也很迷惑，如果说这是树莓派架构不兼容导致的，那之前我mac用 chromium 96, 98, 99 的时候也没出过这问题啊，更何况你的VPS肯定是64位的架构，不存在这个问题

现在我只能猜测是某几个 chromium 版本有点问题，至少最后我推的那个 commit 把我这边的问题解决了，对性能的影响还要再观察一下

**Comment by mark9804** [comment]
我在 dev 分支新增了一个 commit，可能可以修复这个问题，但是我猜副作用是并发性能会非常低，麻烦测试一下

----
我也不知道为什么需要这样，理论上 `["--no-sandbox", "--no-zygote"]` 就足够了，我现在加了一个 `"--single-process"` 才能阻止 
 chromium 继续运行，但是这个 flag 是一个表现很不稳定的 flag ，所以需要做好回滚掉的准备

**Reply by Xm798** [reply]
刚才重启了一下，又复现了。。。checkout 到 dev，好了，真不错

**Reply by mark9804** [reply]
我感觉这是一个概率性问题，概率性问题是最烦的

既然你用 dev 了，顺便观察一下高并发的时候会不会吞图片吧

**Reply by Arondight** [reply]
> 刚才重启了一下，又复现了。。。checkout 到 dev，好了，真不错

有啥问题圈我一下，我晚上把数据库都改了，最好再 pull 一下

**Reply by mark9804** [reply]
> 我晚上把数据库都改了

我来加需求了，tool-db 可以直接操作内存中的数据库吗，改时间戳什么的会方便很多

**Reply by Arondight** [reply]
> 我来加需求了，tool-db 可以直接操作内存中的数据库吗，改时间戳什么的会方便很多

目前的结构不可能，技术上不可行

**Comment by mark9804** [comment]
不能用single process， #680

**Reply by Arondight** [reply]
确实是这样的，要回退掉 https://www.chromium.org/developers/design-documents/process-models/#single-process

**Reply by mark9804** [reply]
但是如果不这么做的话，有什么其他办法解决这个问题呢，这个bug的原因我有两种猜想

1. 浏览器进程在哪里阻塞住了
2. 有僵尸进程

我是从chromium 95.x 更新到98 之后出现的这个问题， Xm 是莫名其妙出现了这个问题，但是表现形式不一样，我能正常绘图，只是一直有个进程在占用CPU让我很不舒服

**Reply by Arondight** [reply]
可以试试这个 https://www.chromium.org/developers/design-documents/process-models/#process-per-site

**Reply by Arondight** [reply]
我觉得在某些 vps 或者小型设备上 cpu 消耗过大可能是渲染进程太多了，按照文档我们这样做只会有一个渲染进程，同时渲染进程不会影响到主进程

**Reply by mark9804** [reply]
> 可以试试这个 https://www.chromium.org/developers/design-documents/process-models/#process-per-site

明天我看一下，感觉per tab或者 per site 看起来会正常一些

**Reply by Arondight** [reply]
我觉得 pertab 首先排除

**Reply by mark9804** [reply]
`--process-per-site` 不能解决这个问题，经过大胆假设我把 `--no-zygote` 干掉了，说不定之前这个 flag 能用是因为 bug 呢

现在重启了两遍看起来挺正常的，不知道是不是这两遍运气好（好？）没碰上，等我下午多试几遍

**Reply by mark9804** [reply]
重启了几遍没问题了，推到 dev @Xm798 看看能不能解决

应该就是 `--no-zygote` 的锅，要么之前行为不正常，要么现在不正常

**Reply by Xm798** [reply]
> 重启了几遍没问题了，推到 dev @Xm798 看看能不能解决
> 
> 应该就是 `--no-zygote` 的锅，要么之前行为不正常，要么现在不正常

更新了，重启几次资源占用正常，图片试了几个目前也正常

**Comment by mark9804** [comment]
又开始了这种随机情况
<img width="531" alt="image" src="https://user-images.githubusercontent.com/9006264/161482078-e1fec5e5-3413-4240-85e9-ec18c6a3f79d.png">

```
Chromium 98.0.4758.106 Built on Debian , running on Debian 11
```
